Invalid coordinates on gesture (and resulting wheel) events inside iframe

aprotyas opened the pull request:
#### c8a66bff2703a2099a9e61c75382cef21ee0c822
<pre>
Invalid coordinates on gesture (and resulting wheel) events inside iframe
<a href="https://bugs.webkit.org/show_bug.cgi?id=266523">https://bugs.webkit.org/show_bug.cgi?id=266523</a>
<a href="https://rdar.apple.com/105243167">rdar://105243167</a>

Reviewed by Wenson Hsieh.

`gesture*` events (and resulting wheel events) originating from an iframe
have incorrect coordinates. Notably, these coordinates are not relative
to the child frame&apos;s origin. This happens because we pass events on to a
child frame&apos;s nodes without accounting for the coordinate space conversion
between the parent frame and the child frame.

This bug is addressed in WebKitAdditions by adopting the &quot;pass event to
subframe&apos;s event handler&quot; pattern in the the gesture event handling code -
as used for other event handling codepaths.

This open source commit adds a layout test in which gesture events are
originated in an embedded frame. The test asserts that the event
coordinates are relative to the child frame, thus providing some
correctness guarantees for the aforementioned fix.

* LayoutTests/fast/events/gesture/resources/gesture-iframe.html: Added.
* LayoutTests/fast/events/gesture/wheel-gesture-event-coordinates-in-iframe-expected.txt: Added.
* LayoutTests/fast/events/gesture/wheel-gesture-event-coordinates-in-iframe.html: Added.

Canonical link: <a href="https://commits.webkit.org/272176@main">https://commits.webkit.org/272176@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/24ca43ebc6b8bc9e932394053c5e6640c740509d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/30812 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/9491 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/32482 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/33311 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/27838 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/11806 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/6737 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/27705 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/31129 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/7969 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/27562 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/6833 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/6981 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/27440 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/34648 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/28042 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/27919 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/33138 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/7028 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/5106 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/30971 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/8745 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/27225 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7290 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/7746 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/7586 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

charliewolfe commented on Source/WebCore/page/EventHandler.h:
```diff
@@ -249,7 +249,7 @@ class EventHandler : public CanMakeCheckedPtr {
 #elif ENABLE(MAC_GESTURE_EVENTS)
     bool dispatchGestureEvent(const PlatformGestureEvent&, const AtomString&, const EventTargetSet&, float, float);
-    WEBCORE_EXPORT bool handleGestureEvent(const PlatformGestureEvent&);
+    WEBCORE_EXPORT HandleUserInputEventResult handleGestureEvent(const PlatformGestureEvent&);
```
We’ll need to do this eventually (so maybe it’s fine), but there’s no reason for this to change in this PR. The plumbing isn’t in place for `RemoteUserInputEventData` to be sent to the UI process for gesture events, so we only need the bool.

aprotyas replied:
Yes I agree! I just made the change to resemble the other event handling methods, but I agree we don't need this interface to change. I'll remove this change for now.

aprotyas replied:
Addressed in https://github.com/WebKit/WebKit/commit/24ca43ebc6b8bc9e932394053c5e6640c740509d

aprotyas committed:
Invalid coordinates on gesture (and resulting wheel) events inside iframe
https://bugs.webkit.org/show_bug.cgi?id=266523
rdar://105243167

Reviewed by Wenson Hsieh.

`gesture*` events (and resulting wheel events) originating from an iframe
have incorrect coordinates. Notably, these coordinates are not relative
to the child frame's origin. This happens because we pass events on to a
child frame's nodes without accounting for the coordinate space conversion
between the parent frame and the child frame.

This bug is addressed in WebKitAdditions by adopting the "pass event to
subframe's event handler" pattern in the the gesture event handling code -
as used for other event handling codepaths.

This open source commit adds a layout test in which gesture events are
originated in an embedded frame. The test asserts that the event
coordinates are relative to the child frame, thus providing some
correctness guarantees for the aforementioned fix.

* LayoutTests/fast/events/gesture/resources/gesture-iframe.html: Added.
* LayoutTests/fast/events/gesture/wheel-gesture-event-coordinates-in-iframe-expected.txt: Added.
* LayoutTests/fast/events/gesture/wheel-gesture-event-coordinates-in-iframe.html: Added.

Canonical link: https://commits.webkit.org/272176@main
